### ISSUE TITLE
Unlimited P2P Command Size Enables Mem Crash DoS Attack

### ISSUE THREAD

**Issue by xhyumiracle** [QuarkChain/pyquarkchain#399]
P2P_COMMAND_SIZE_LIMIT is set to 2^32-1, which is too large for command packet. This enables attackers to send data packet as large as 4G each time to any master/slave node, which may crash the node server by occupying the memory. 
Moreover, due to the design of cluster, the whole cluster would shutdown if any master/slave subprocess exit.
```python
P2P_COMMAND_SIZE_LIMIT = (2 ** 32) - 1  # unlimited right now
```
in protocol.py:
```python
async def read_metadata_and_raw_data(self):
        ...
        if size > self.env.quark_chain_config.P2P_COMMAND_SIZE_LIMIT:
            raise RuntimeError("{}: command package exceed limit".format(self.name))
        ...
```

**Comment by xhyumiracle** [comment]
BTW, here is where the read happens:
```python
raw_data_without_size = await self.__read_fully(1 + 8 + size)
```
And __read_fully keeps extending the bytearray until the size reaches n.
```python
    async def __read_fully(self, n, allow_eof=False):
        ba = bytearray()
        bs = await self.reader.read(n)
        if allow_eof and len(bs) == 0 and self.reader.at_eof():
            return None

        ba.extend(bs)
        while len(ba) < n:
            bs = await self.reader.read(n - len(ba))
            if len(bs) == 0 and self.reader.at_eof():
                raise RuntimeError("{}: read unexpected EOF".format(self.name))
            ba.extend(bs)
        return ba
```

**Comment by qizhou** [comment]
This is an interesting finding.  Any suggestions from Bitcoin core or Ethereum code?  I know Bitcoin is restricting block size to be up to 1M, but glad to know if you can pinpoint the actual code is.  Thanks.

**Comment by qizhou** [comment]
Bitcoin's command size is 4M (and max serialization size is 32M).  We should adopt a similar number (may be a bit larger), and limit the command size for batch download such as GET_ROOT_BLOCK_LIST_REQUEST/GET_ROOT_BLOCK_LIST_RESPONSE and GET_MINOR_BLOCK_HEADER_LIST_REQUEST/GET_MINOR_BLOCK_HEADER_LIST_RESPONSE

**Comment by qizhou** [comment]
Ethereum's frame size is up to 16MB

https://github.com/ethereum/go-ethereum/blob/358535188852bd9d8e351ccbb0d1fa608284a77b/p2p/rlpx.go#L657

BCH ABC's command size is up to 32MB, and BCH SV's command size is up to 128MB (for block like commands)

**Comment by qizhou** [comment]
We also need to measure the max size of a minor block and root block and determine the max size of batch downloads.

**Comment by qizhou** [comment]
This should be fixed by #431
